tests/system_tests_tcp_adaptor_tls.py

```diff
@@ -30,15 +30,101 @@ def test_authenticate_peer(self):
+                         ncat_port=self.authenticate_peer_port,
+                         use_ssl=True,
+                         use_client_cert=False,
+                         error_ok=True)
+        # Look for a log line that proves the peer did not return a certificate.
+        self.INTA.wait_log_message("peer did not return a certificate")
+        self.logger.log("TCP_TEST Stop %s SUCCESS" % name)
+
+    def test_connector_bad_server_cert(self):
+        if not ncat_available():
+            self.skipTest("Ncat utility is not available")
+        name = "test_connector_bad_server_cert"
+        self.logger.log("TCP_TEST TLS Start %s" % name)
+
+        # The router presents a cert to the echo server but the echo server is
+        # responding back with a different ca cert that the router cannot verify.
+        # Hence the router side TLS fails certificate verification.
+        self.ncat_runner(name,
+                         client="INTA",
+                         server="INTA",
+                         logger=self.logger,
+                         ncat_port=self.bad_server_port,
+                         use_ssl=True,
+                         use_client_cert=False,
+                         error_ok=True)
+        # Look for a log line that proves the certificate verification failure.
+        self.INTA.wait_log_message("certificate verify failed")
+        self.logger.log("TCP_TEST Stop %s SUCCESS" % name)
+
+    def test_xxx_present_bad_cert_to_good_listener_good_connector(self):
+        if not ncat_available():
+            self.skipTest("Ncat utility is not available")
+        name = "test_xxx_present_bad_cert_to_good_listener_good_connector"
+        self.logger.log("TCP_TEST TLS Start %s" % name)
+
+        # Run the ncat command using a bad ca certificate to a tcpListener that is good
+        # and a tcpConnector that is good. The router must reject the connection made by
+        # ncat since it presents an unknown certificate.
+        self.ncat_runner(name,
+                         client="INTA",
+                         server="INTC",
+                         logger=self.logger,
+                         ncat_port=self.good_server_port,
+                         use_ssl=True,
+                         use_client_cert=False,
+                         use_bad_ca_cert=True,
+                         error_ok=True)
+        # Look for a log line that says "certificate unknown"
+        self.INTA.wait_log_message("certificate unknown")
+        self.logger.log("TCP_TEST Stop %s SUCCESS" % name)
+
+    def test_yyy_present_bad_cert_to_good_listener_bad_connector(self):
+        if not ncat_available():
+            self.skipTest("Ncat utility is not available")
+        name = "test_yyy_present_bad_cert_to_good_listener_bad_connector"
+        self.logger.log("TCP_TEST TLS Start %s" % name)
+
+        # Run the ncat command using a bad ca certificate to a tcpListener that is good
```

Comment:
I haven't yet looked at full log messages from this, but I want to just ask @ganeshmurthy if you think that https://issues.apache.org/jira/browse/PROTON-1870 is implemented, that is, we get good meaningful errors from TLS failures from Proton